pkg/virtctl/console/console.go

```diff
@@ -93,19 +93,53 @@ func (c *Console) Run(cmd *cobra.Command, args []string) error {
 	writeStop := make(chan error)
 	readStop := make(chan error)
 
+	// Wait until the virtual machine is in running phase, user interrupt or timeout
+	runningChan := make(chan error)
+	timeoutChan := time.Tick(time.Duration(timeout) * time.Minute)
+	waitInterrupt := make(chan os.Signal, 1)
+	signal.Notify(waitInterrupt, os.Interrupt)
+
 	go func() {
 		con, err := virtCli.VirtualMachineInstance(namespace).SerialConsole(vmi)
-		if err != nil {
-			resChan <- err
-			return
+		if strings.Contains(fmt.Sprint(err), "not found") {
+			runningChan <- err
+		}
+
+		for err != nil {
+			// Sleep to prevent denial of service on the api server
+			time.Sleep(500 * time.Millisecond)
+			con, err = virtCli.VirtualMachineInstance(namespace).SerialConsole(vmi)
 		}
 
+		runningChan <- nil
 		resChan <- con.Stream(kubecli.StreamOptions{
 			In:  stdinReader,
 			Out: stdoutWriter,
 		})
 	}()
 
+	select {
+	case <-timeoutChan:
+		return fmt.Errorf("Timeout trying to connect to the virtual machine instance")
```

Comment:
there is a chance now that a timeout happens while the goroutine above actually succeeds to connect. Is there anything to clean up in this particular case? Particularly, doesn't con.Stream hold a connection state that would be better off closed with explicit close() call?

Reply:
Because this is a command line program if the timeout or the interrupt channel is active the program will exit so the connection will be closed anyway.

Reply:
Do you mean some deferred callback handles it for us? If so, great.